src/target/riscv/riscv-013.c

```diff
@@ -2945,17 +2970,74 @@ static int riscv013_halt_current_hart(struct target *target)
 	return ERROR_OK;
 }
 
-static int riscv013_resume_current_hart(struct target *target)
+/* Select all harts that were prepped and that are selectable, clearing the
+ * prepped flag on the harts that actually were selected. */
+static int select_prepped_harts(struct target *target, bool *use_hasel)
+{
+	dm013_info_t *dm = get_dm(target);
+	if (!dm->hasel_supported) {
+		RISCV_INFO(r);
+		r->prepped = false;
+		*use_hasel = false;
+		return ERROR_OK;
+	}
+
+	assert(dm->hart_count);
+	unsigned hawindow_count = (dm->hart_count + 31) / 32;
+	uint32_t hawindow[hawindow_count];
+
+	memset(hawindow, 0, sizeof(uint32_t) * hawindow_count);
+
+	target_list_t *entry;
+	unsigned total_selected = 0;
+	list_for_each_entry(entry, &dm->target_list, list) {
+		struct target *t = entry->target;
+		riscv_info_t *r = riscv_info(t);
+		riscv013_info_t *info = get_info(t);
+		unsigned index = info->index;
+		LOG_DEBUG("index=%d, coreid=%d, prepped=%d", index, t->coreid, r->prepped);
+		if (r->prepped) {
+			hawindow[index / 32] |= 1 << (index % 32);
+			r->prepped = false;
+			total_selected++;
+		}
+		index++;
+	}
+
+	/* Don't use hasel if we only need to talk to one hart. */
+	if (total_selected <= 1) {
+		*use_hasel = false;
+		return ERROR_OK;
+	}
+
+	for (unsigned i = 0; i < hawindow_count; i++) {
+		if (dmi_write(target, DMI_HAWINDOWSEL, 0) != ERROR_OK)
```

Comment:
The code seems to be writing hawindowsel to zero on each iteration.  Is that correct, or is it supposed to be writing it to the value of the variable "i" on each iteration?

Reply:
Yes! Good catch.